File: ir_codegen/irGenerator.js
function generateIR(ast) {
    const ir = [];
  
    function walk(node) {
      switch (node.type) {
        case "Program":
          node.body.forEach(walk);
          break;
  
        case "VariableDeclaration":
          for (const decl of node.declarations) {
            ir.push({
              op: "assign",
              target: decl.id.name,
              value: extractExpr(decl.init)
            });
          }
          break;
  
        case "ExpressionStatement":
          walk(node.expression);
          break;
  
        case "AssignmentExpression":
          ir.push({
            op: "assign",
            target: node.left.name,
            value: extractExpr(node.right)
          });
          break;
  
        case "BinaryExpression":
          return {
            op: node.operator,
            left: extractExpr(node.left),
            right: extractExpr(node.right)
          };
  
        case "IfStatement":
          ir.push({ op: "if", test: extractExpr(node.test) });
          walk(node.consequent);
          if (node.alternate) {
            ir.push({ op: "else" });
            walk(node.alternate);
          }
          ir.push({ op: "endif" });
          break;
  
        case "WhileStatement":
          ir.push({ op: "while", test: extractExpr(node.test) });
          walk(node.body);
          ir.push({ op: "endwhile" });
          break;
  
        case "ForStatement":
          walk(node.init);
          ir.push({ op: "for", test: extractExpr(node.test), update: extractExpr(node.update) });
          walk(node.body);
          ir.push({ op: "endfor" });
          break;
  
        case "FunctionDeclaration":
          ir.push({
            op: "function",
            name: node.id.name,
            params: node.params.map(p => p.name)
          });
          walk(node.body);
          ir.push({ op: "endfunction" });
          break;
  
        case "CallExpression":
          ir.push({
            op: "call",
            callee: node.callee.name,
            args: node.arguments.map(extractExpr)
          });
          break;
  
        case "ReturnStatement":
          ir.push({
            op: "return",
            value: extractExpr(node.argument)
          });
          break;
  
        case "TryStatement":
          ir.push({ op: "try" });
          walk(node.block);
          ir.push({ op: "catch", param: node.handler.param.name });
          walk(node.handler.body);
          ir.push({ op: "endtry" });
          break;
  
        case "BlockStatement":
          node.body.forEach(walk);
          break;
  
        default:
          console.warn("Unhandled node type:", node.type);
      }
    }
  
    function extractExpr(expr) {
      if (!expr) return null;
  
      switch (expr.type) {
        case "NumericLiteral":
        case "StringLiteral":
          return expr.value;
  
        case "Identifier":
          return expr.name;
  
        case "BinaryExpression":
          return {
            op: expr.operator,
            left: extractExpr(expr.left),
            right: extractExpr(expr.right)
          };
  
        case "CallExpression":
          return {
            op: "call",
            callee: expr.callee.name,
            args: expr.arguments.map(extractExpr)
          };
  
        case "ArrayExpression":
          return expr.elements.map(extractExpr);
  
        case "ObjectExpression":
          return Object.fromEntries(
            expr.properties.map(prop => [
              prop.key.name || prop.key.value,
              extractExpr(prop.value)
            ])
          );
  
        default:
          return "<unsupported>";
      }
    }
  
    walk(ast);
    return ir;
  }
  
  module.exports = { generateIR };
  

File: backend/compiler/ir_generator.py
def generate_ir(node):
    if node['type'] == 'Program':
        return [generate_ir(stmt) for stmt in node['body']]

    if node['type'] == 'VariableDeclaration':
        init = generate_ir(node['init']) if node['init'] else None
        return {'type': 'VarDecl', 'name': node['id'], 'init': init}

    if node['type'] == 'Literal':
        return node['value']

    if node['type'] == 'Identifier':
        return node['name']

    if node['type'] == 'BinaryExpression':
        left = generate_ir(node['left'])
        right = generate_ir(node['right'])
        return {'type': 'BinOp', 'operator': node['operator'], 'left': left, 'right': right}

    if node['type'] == 'ExpressionStatement':
        return generate_ir(node['expression'])

    if node['type'] == 'CallExpression':
        args = [generate_ir(arg) for arg in node['arguments']]
        return {'type': 'Call', 'callee': node['callee'], 'arguments': args}

    if node['type'] == 'IfStatement':
        test = generate_ir(node['test'])
        consequent = [generate_ir(stmt) for stmt in node['consequent']]
        alternate = [generate_ir(stmt) for stmt in node['alternate']] if node['alternate'] else None
        return {'type': 'If', 'test': test, 'consequent': consequent, 'alternate': alternate}

    if node['type'] == 'ForStatement':
        init = generate_ir(node['init'])
        test = generate_ir(node['test'])
        update = generate_ir(node['update'])
        body = [generate_ir(stmt) for stmt in node['body']]
        return {'type': 'For', 'init': init, 'test': test, 'update': update, 'body': body}

    if node['type'] == 'WhileStatement':
        test = generate_ir(node['test'])
        body = [generate_ir(stmt) for stmt in node['body']]
        return {'type': 'While', 'test': test, 'body': body}

    if node['type'] == 'FunctionDeclaration':
        body = [generate_ir(stmt) for stmt in node['body']]
        return {'type': 'Function', 'name': node['name'], 'params': node['params'], 'body': body}

    if node['type'] == 'ReturnStatement':
        arg = generate_ir(node['argument']) if node['argument'] else None
        return {'type': 'Return', 'argument': arg}

    raise Exception(f"Unsupported AST node {node['type']}")
